kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,915 @@
+			return nil
+		}
+	}
+
+	loadBalancerIPs, ipv4pools, ipv6pools, err := c.parseAnnotations(svc.Annotations)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to parse annotations for service %s/%s", svc.Namespace, svc.Name)
+		return err
+	}
+
+	// Calico assigned IP previously, no longer managed by us, release IPs assigned by calico
+	if !IsCalicoManagedLoadBalancer(svc, c.cfg.AssignIPs) {
+		for ip := range c.allocationTracker.ipsByService[svcObj.namespacedName] {
+			err = c.releaseIP(svcObj.namespacedName, ip)
+			if err != nil {
+				log.WithError(err).Errorf("Failed to release IP for %s/%s", svc.Namespace, svc.Name)
+				continue
+			}
+			c.allocationTracker.releaseAddressFromService(svcObj.namespacedName, ip)
+		}
+	}
+
+	if loadBalancerIPs != nil {
+		// Check that service has assigned IPs to the once specified in annotations
+		lbIPs := make(map[string]bool)
+		for _, ip := range loadBalancerIPs {
+			lbIPs[ip.String()] = true
+		}
+		for ip := range c.allocationTracker.ipsByService[svcObj.namespacedName] {
+			if _, ok := lbIPs[ip]; !ok {
+				err = c.releaseIP(svcObj.namespacedName, ip)
+				if err != nil {
+					log.WithError(err).Errorf("Failed to release IP for %s/%s", svc.Namespace, svc.Name)
+					return err
+				}
+				c.allocationTracker.releaseAddressFromService(svcObj.namespacedName, ip)
+			}
+		}
+	} else if ipv4pools != nil || ipv6pools != nil {
+		// If pool annotations are specified, we need to check that the IPs assigned are from the specified pools
+		for ip := range c.allocationTracker.ipsByService[svcObj.namespacedName] {
+			if !poolContains(ip, ipv4pools) && !poolContains(ip, ipv6pools) {
+				err = c.releaseIP(svcObj.namespacedName, ip)
+				if err != nil {
+					log.WithError(err).Errorf("Failed to release IP for %s/%s", svc.Namespace, svc.Name)
+					return err
+				}
+				c.allocationTracker.releaseAddressFromService(svcObj.namespacedName, ip)
+			}
+		}
+	} else {
+		// No annotations are specified, check that the IPs assigned aren't from Manual pool from earlier assignment
+		for ip := range c.allocationTracker.ipsByService[svcObj.namespacedName] {
+			pool, err := c.poolForIP(ip)
+			if err != nil {
+				return err
+			}
+			if pool != nil {
+				if pool.Spec.AssignmentMode == api.Manual {
+					err = c.releaseIP(svcObj.namespacedName, ip)
+					if err != nil {
+						log.WithError(err).Errorf("Failed to release IP for %s/%s", svc.Namespace, svc.Name)
+						return err
+					}
+					c.allocationTracker.releaseAddressFromService(svcObj.namespacedName, ip)
+				}
+			}
+		}
+	}
+
+	if IsCalicoManagedLoadBalancer(svc, c.cfg.AssignIPs) {
+		statusIPs = make(map[string]bool)
+		for _, ingress := range svc.Status.LoadBalancer.Ingress {
+			statusIPs[ingress.IP] = true
+		}
+
+		for ip := range c.allocationTracker.ipsByService[svcObj.namespacedName] {
+			if _, ok := statusIPs[ip]; !ok {
+				break
+			}
+			return nil
+		}
+
+		_, err = c.assignIP(svc)
```

Comment:
Won't we hit this even if IPs were allocated based on annotations above? I think the structuring of this logic is a bit tricky to navigate (for me at least), as you need to track the path of each service all the way through a fairly large and complicated function.

It might make sense to decompose this a bit into helper functions - e.g.,

```go
func assignFromAnnotations(svc *ServiceObject) error { }
func assignAutomatically(svc *ServiceObject) error { }
func updateStatusIfNeeded(svc *ServiceObject) error { }
```